MaxText/standalone_dataloader.py

```diff
@@ -0,0 +1,77 @@
+"""
+ Copyright 2023 Google LLC
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ """
+
+# pylint: disable=g-bad-todo, abstract-method, consider-using-with, ungrouped-imports
+""" Standalone data loader - only loads data for each training step, accesses storage needs."""
+
+# Calling jax.device_count here prevents a "TPU platform already registered" error.
+# See github.com/google/maxtext/issues/20 for more
+import jax
+import os
+
+from typing import Sequence
+import datetime
+from absl import app
+import numpy as np
+
+import pyconfig
+from train import validate_train_config, get_first_step, load_next_batch, setup_train_loop
+
+from jax.experimental.compilation_cache import compilation_cache as cc
+
+
+def data_load_loop(config, state=None):
+  """Main data loader loop.
+    Loads batches of data for each training step.
+  """
+  _, _, _, _, _, _, _, data_iterator, state = setup_train_loop(config)
+
+  example_batch = None
+
+  start = datetime.datetime.now()
+  start_step = get_first_step(state)
+
+  # Actual data loading steps
+  for step in np.arange(start_step, config.steps):
+    example_batch = load_next_batch(data_iterator, example_batch, config)
+    jax.block_until_ready(example_batch)
+    if step==0:
+      new_time = datetime.datetime.now()
+      print("First step completed in ", new_time-start," seconds")
+
+  end = datetime.datetime.now()
+  print(config.steps," batches loaded in ", end-start ," seconds, on host ", jax.process_index())
```

Comment:
Mostly a nit but will leave the code a bit clearer.

What are you trying to time?

I think you're trying to time separately:
(a) the time to first batch
(b) once initialized the time to load N batches.

If that case, you need should:
* time (a) outside of the for loop (and end the timer on (a) when `block_until_ready`)
* time (b) by starting the timer, loading n batches and `block_until_ready` on all of them at the end.

I actually don't think it matters much in this case that you have block_until_ready's in the for loop, but you're adding a curious dependency between the data.

Reply:
Thanks for the comment!
I am trying to time separately:

1. The time to load the first batch 
Is your suggestion to do the following outside of the for loop?
```
start = datetime.datetime.now()
start_step = get_first_step(state)
jax.block_until_ready(start_step)
first_end = datetime.datetime.now()
time_to_load_first_batch = first_end-start
```
In this case, I think the iterator has not moved, only jax spmd initialization has happened.

2. Time to load n batches, including the first one. In xpk runs, I was able to see the time taken to load the batches on the slowest host at the end.

```
start = datetime.datetime.now()
for step in np.arange(start_step, config.steps):
    example_batch = load_next_batch(data_iterator, example_batch, config)
jax.block_until_ready(example_batch) # waiting for the last batch
end = datetime.datetime.now()
```

Do you think adding ` jax.block_until_ready(example_batch)` outside of the for loop,  serves my intent of waiting until last batch is loaded?

Reply:
No, in 1 we should start when the program starts and include a 
```
example_batch = load_next_batch(data_iterator, example_batch, config)
```

Yes 2 is fine.